### ISSUE TITLE
Implicit conversion from contract to address

### ISSUE THREAD

**Issue by frangio** [ethereum/solidity#12328]
If a contract wants to query its own balance of some (ERC20) token, the expression is `token.balanceOf(address(this))`. As far as I can tell, there is no reason to force a manual cast to address. It would be simpler to write, but more importantly simpler to read and understand, if the expression was `token.balanceOf(this)`.

Couldn't find an existing issue about this conversion in particular, and it seems like it is not a breaking change so it's worth discussing on its own.

I'm not sure if this became a requirement in 0.5.0. There is this item in the changelog:
>Explicitly convert values of contract type to addresses before using an address member. Example: if c is a contract, change c.transfer(...) to address(c).transfer(...).

But it applies to members only. I can see that that's more complicated. But in argument position I think this should be an implicit cast.

**Comment by cameel** [comment]
The fact that contract types are no longer implicitly convertible to `address` was a breaking change added in #4321 (0.5.0). Looks like it was not explicitly mentioned in the changelog or [Solidity v0.5.0 Breaking Changes > Explicitness Requirements](https://docs.soliditylang.org/en/latest/050-breaking-changes.html#explicitness-requirements) and instead was assumed to be implied by this:
> - Explicit conversions between unrelated contract types are now disallowed. You can only convert from a contract type to one of its base or ancestor types. If you are sure that a contract is compatible with the contract type you want to convert to, although it does not inherit from it, you can work around this by converting to `address` first. Example: if `A` and `B` are contract types, `B` does not inherit from `A` and `b` is a contract of type `B`, you can still convert `b` to type `A` using `A(address(b))`. Note that you still need to watch out for matching payable fallback functions, as explained below.

If an implicit conversion was allowed, you could still do `A(b)` because `b` would be implicitly converted to `address` and then the address explicitly converted to `A`.

**Comment by frangio** [comment]
Thanks for the context! I see the reasoning behind that, but is it possible to disallow `A(b)` while allowing implicit conversion to address in other contexts?

My position is that safe conversions should be implicit as much as possible. Otherwise, they dillute the attention given to unsafe casts because the syntax is indistinguishable.

**Comment by cameel** [comment]
I think it could be done by just treating it as a special case and disallowing it in contract conversions. If it's really only that one case then I think it would not be a bad idea and I don't see much harm it automatically converting a contract to an address. I'll add this to the design backlog and we'll discuss it on a call.

The only thing is that it goes against #11700, which we have decided to implement some time ago. That issue would actually add more cases where an explicit `address` conversion is necessary. The main concern in that issue was that comparison between contracts could be mistaken for an operation that compares more than just an address, e.g. checks if it's actually an instance of the same contract type and comparison of addresses is always clear (and sometimes even more readable).

**Comment by k06a** [comment]
Implicit contract/interface conversion to `address` would create troubles with methods signatures: `address .transfer(uint256)` and `address.send(uint256)` and others. We could extract this methods away from address: `evm.transfer(address, uint256)` or smith like this. Casting to `address` in arguments feels really bad.

**Comment by chriseth** [comment]
> Implicit contract/interface conversion to `address` would create troubles with methods signatures: `address .transfer(uint256)` and `address.send(uint256)` and others. We could extract this methods away from address: `evm.transfer(address, uint256)` or smith like this. Casting to `address` in arguments feels really bad.

Implicit conversion is not performed for member lookup, so I don't see a problem here.

**Comment by k06a** [comment]
@chriseth I would say `address` is not base contract/interface, that’s why we don’t have this collision. But if we would implicitly cast to address we will definitely have this issue.

**Comment by adhusson** [comment]
It looks like losing contract->address implicit conversion was a side-effect of disallowing sideways contract casting. Sadly that represents quite a loss in quality of life. I don't see any good reason for it, you can have both.

edited for clarity/fixes: this responds to

> If an implicit conversion was allowed, you could still do A(b) because b would be implicitly converted to address and then the address explicitly converted to A.

Since [0.8.0](https://blog.soliditylang.org/2020/12/16/solidity-v0.8.0-release-announcement/) explicit conversions "are only possible when there is at most one change in sign, width, payability or type-category". You could add "and none in contract hierarchy branch" to the previous sentence. Then you take `address` as the root of the contract inheritance tree, and you can reenable implicit conversion everywhere. Then you would get the following (comparing with uints):

|              | numbers    | contracts    |
|--------------|------------|--------------|
| value        | x:uint24   | x:Contract1  |
| OK function  | f(uint256) | f(address)   |
| Bad function | g(bytes3)  | f(Contract2) |
| f(x)         | ✅          | ✅            |
| g(x)         | ❌          | ❌            |

(`Contract2` neither an ancestor nor a descendant of `Contract`.)

**Comment by clouds56** [comment]
Shall we reactivate this?
I'm writing interface for my contract, and I have to writing `address(token)` every where.
I really want to annotate `token` with `IERC20` since it indeed is, but at the same time it would be used as address (more often than used as `IERC20`)
```solidity
contract MyContract {
  IERC20 token;
}
```

Even in uniswap v3, where interface are well used, they have to use `address` everywhere in contracts
https://github.com/Uniswap/v3-core/blob/d8b1c635c275d2a9450bd6a78f3fa2484fef73eb/contracts/UniswapV3Pool.sol#L41-L46
```solidity
    /// @inheritdoc IUniswapV3PoolImmutables
    address public immutable override factory;
    /// @inheritdoc IUniswapV3PoolImmutables
    address public immutable override token0;
    /// @inheritdoc IUniswapV3PoolImmutables
    address public immutable override token1;
```
or AAVE https://github.com/aave/aave-v3-core/blob/29ff9b9f89af7cd8255231bc5faf26c3ce0fb7ce/contracts/protocol/tokenization/AToken.sol#L33-L34
```solidity
  address internal _treasury;
  address internal _underlyingAsset;
```
and used like this
```solidity
IERC20(_underlyingAsset).safeTransfer(receiverOfUnderlying, amount);
```

I personally think this pattern is not perfect since every time you explicit cast you might convert the address to wrong type like
```solidity
IERC721(_underlyingAsset).transfer(receiverOfUnderlying, amount);
```
and compiler couldn't find out for me every time.

If I have token `IERC20` at first place, these kind of conversation would be blocked by compiler
```solidity
IERC20 token;
IERC721(token).transfer(to, amount);
```

The situation now is in ecosystem every one uses and passes `address`, and once you uses `IERC20` to tell the underlying type, you have write `address(token)` every where.
If we have the implicit conversation, it might encourage others use interface instead of `address`, since it could also be used as address.
If we have hard choice on when the implicit conversations should be make, just limit to pass it as parameters when calling some function.

TL; DR now we still in dangerous position due to `address` everywhere, and I think it is related to no implicit conversation.
